src/rewrites.rs

```diff
@@ -849,27 +1037,519 @@ impl MultiPatterns {
+    }
+}
+
+/// Resolve cycle by adding node to blacklist
+///
+/// # Parameters
+///
+/// - `egraph`: E-graph of interest
+/// - `cycle`: list of nodes within the cycle
+/// - `added_node_to_order`: HashMap, map from node to the order that it was added into the egraph
+fn resolve_cycle(
+    egraph: &mut EGraph<Mdl, TensorAnalysis>,
+    cycle: &[Mdl],
+    added_node_to_order: &HashMap<Mdl, usize>,
+) {
+    // Check if any node in cycle is already in blacklist
+    let already_solved = cycle
+        .iter()
+        .any(|node| egraph.analysis.blacklist_nodes.contains(node));
+    if !already_solved {
+        assert!(cycle.len() > 0);
+        let (ord, n) = cycle
+            .iter()
+            .map(|node| {
+                let order = added_node_to_order
+                    .get(node)
+                    .map_or(-1, |index| *index as i32);
+                (order, node.clone())
+            })
+            .max_by_key(|(o, _)| *o)
+            .unwrap();
+        assert!(ord >= 0);
+        egraph.analysis.blacklist_nodes.insert(n.clone());
+    }
+}
+
+/// Traverse the EGraph and get a set of cycles (reachable from root)
+///
+/// # Parameters
+///
+/// - `egraph`: E-graph of interest
+/// - `root`: Id of root eclass
+/// - `paths_from_root`: HashMap storing for each eclass, one path from the root eclass. The path
+///         is represented by a list of (eclass, enode) that composes the path
+/// - `cycles`: list of cycles. Each cycle is a list of nodes. A cycle of 1->2->4->3->1 will be
+///         stored as [1,2,4,3]
+fn get_cycles(
+    egraph: &EGraph<Mdl, TensorAnalysis>,
+    root: Id,
+    paths_from_root: &mut HashMap<Id, Vec<(Id, Mdl)>>,
+    cycles: &mut Vec<Vec<Mdl>>,
+) {
+    // Get a map from Id to the eclass objects, since egg doesn't provide accessing eclass from Id
+    let id_to_class: HashMap<Id, &EClass<Mdl, ValTnsr>> =
+        egraph.classes().map(|class| (class.id, class)).collect();
+
+    get_cycles_rec(
+        egraph,
+        root,
+        /*path_to_here=*/ Vec::<(Id, Mdl)>::new(),
+        &id_to_class,
+        paths_from_root,
+        cycles,
+    );
+}
+
+/// Traverse the EGraph in DFS order, update paths_from_root and cycles on the fly
+///
+/// # Parameters
+///
+/// - `egraph`: E-graph of interest
+/// - `eclass`: The current eclass that we are visiting
+/// - `path_to_here`: A path from root to this eclass
+/// - `id_to_class`: Map from eclass ID to the eclass objects
+/// - `paths_from_root`: HashMap storing for each eclass, one path from the root eclass. The path
+///         is represented by a list of (eclass, enode) that composes the path
+/// - `cycles`: list of cycles. Each cycle is a list of nodes. A cycle of 1->2->4->3->1 will be
+///         stored as [1,2,4,3]
+fn get_cycles_rec(
+    egraph: &EGraph<Mdl, TensorAnalysis>,
+    eclass: Id,
+    path_to_here: Vec<(Id, Mdl)>,
+    id_to_class: &HashMap<Id, &EClass<Mdl, ValTnsr>>,
+    paths_from_root: &mut HashMap<Id, Vec<(Id, Mdl)>>,
```

Comment:
Looks like you're using paths_from_root as a set not map. Do you really need a map here?

Reply:
Changed to HashSet